analytic_engine/src/sst/parquet/async_reader.rs

```diff
@@ -143,72 +145,80 @@ impl<'a> Reader<'a> {
             return Ok(());
         }
 
-        let sst_meta = Self::read_sst_meta(self.storage, self.path, &self.meta_cache).await?;
+        let meta_data = Self::read_sst_meta(self.storage, self.path, &self.meta_cache).await?;
 
         let row_projector = self
             .projected_schema
-            .try_project_with_key(&sst_meta.schema)
+            .try_project_with_key(&meta_data.custom.schema)
             .map_err(|e| Box::new(e) as _)
             .context(Projection)?;
-        self.meta_data = Some(sst_meta);
+        self.meta_data = Some(meta_data);
         self.row_projector = Some(row_projector);
         Ok(())
     }
 
+    async fn load_meta_data_from_storage(
+        storage: &ObjectStoreRef,
+        object_meta: &ObjectMeta,
+    ) -> Result<ParquetMetaDataRef> {
+        let meta_data =
+            file_format::parquet::fetch_parquet_metadata(storage.as_ref(), object_meta, None)
+                .await
+                .map_err(|e| Box::new(e) as _)
+                .context(DecodeSstMeta)?;
+        Ok(Arc::new(meta_data))
+    }
+
     async fn read_sst_meta(
         storage: &ObjectStoreRef,
         path: &Path,
         meta_cache: &Option<MetaCacheRef>,
-    ) -> Result<SstMetaData> {
-        let object_meta = storage.head(path).await.context(ObjectStoreError {})?;
-        let file_size = object_meta.size;
-
-        let get_metadata_from_storage = || async {
-            let mut reader = CachableParquetFileReader::new(
-                storage.clone(),
-                path.clone(),
-                None,
-                Some(file_size),
-            );
-            let metadata = reader.get_metadata().await.context(ParquetError {})?;
-
-            if let Some(cache) = meta_cache {
-                cache.put(path.to_string(), metadata.clone());
+    ) -> Result<MetaData> {
+        if let Some(cache) = meta_cache {
+            if let Some(meta_data) = cache.get(path.as_ref()) {
+                return Ok(meta_data);
             }
-            Ok(metadata)
-        };
-
-        let metadata = if let Some(cache) = meta_cache {
-            if let Some(cached_data) = cache.get(path.as_ref()) {
-                cached_data
-            } else {
-                get_metadata_from_storage().await?
+        }
+        // The metadata can't be found in the cache, and let's fetch it from the
+        // storage.
+
+        let meta_data = {
+            let object_meta = storage.head(path).await.context(ObjectStoreError {})?;
+            let parquet_meta_data =
+                Self::load_meta_data_from_storage(storage, &object_meta).await?;
+            let kv_metas = parquet_meta_data
```

Comment:
It seems this KV is saved twice, is there any possible to reduce this?

Reply:
It's true. Maybe we can reset the extended kv metas in parquet metadata, because they are already decoded in the custom metadata.

How do you like such reset?

Reply:
Yeah, exteneded KV is mainly used by ourselves, so we can erase them before saving to cache.

Reply:
I've fixed this already. PTAL again.